src/theory/quantifiers/extended_rewrite.cpp

```diff
@@ -1563,6 +1585,571 @@ Node ExtendedRewriter::extendedRewriteArith(Node ret)
+    {
+      return true;
+    }
+  }
+  return false;
+}
+
+void ExtendedRewriter::spliceBv(Node a,
+                                Node b,
+                                std::vector<Node>& av,
+                                std::vector<Node>& bv)
+{
+  Assert(a.getType() == b.getType());
+  if (a.getKind() == BITVECTOR_CONCAT || b.getKind() == BITVECTOR_CONCAT)
+  {
+    unsigned counter = bv::utils::getSize(a) - 1;
+    bool reca = a.getKind() == BITVECTOR_CONCAT;
+    Node cterm = reca ? a : b;
+    Node oterm = reca ? b : a;
+    for (const Node& ctermc : cterm)
+    {
+      // extract relevant portion of other child
+      unsigned csize = bv::utils::getSize(ctermc);
+      Assert((counter + 1) >= csize);
+      Node extract =
+          bv::utils::mkExtract(oterm, counter, (counter + 1) - csize);
+      extract = Rewriter::rewrite(extract);
+      // recurse
+      Node c1 = reca ? ctermc : extract;
+      Node c2 = reca ? extract : ctermc, strict;
+      spliceBv(c1, c2, av, bv);
+      // decrement counter
+      if (counter >= csize)
+      {
+        counter = counter - csize;
+      }
+    }
+  }
+  else
+  {
+    av.push_back(a);
+    bv.push_back(b);
+  }
+}
+
+int ExtendedRewriter::spliceBvConstBit(Node n1, Node n2, std::vector<Node>& nv)
+{
+  if (n1 == n2)
+  {
+    return -1;
+  }
+  Trace("q-ext-rewrite-debug")
+      << "Splice constant bv bit " << n1 << " " << n2 << std::endl;
+  // splice the children
+  std::vector<Node> rc1;
+  std::vector<Node> rc2;
+  spliceBv(n1, n2, rc1, rc2);
+  Assert(rc1.size() == rc2.size());
+  int diff_index = -1;
+  for (unsigned r = 0; r < rc1.size(); r++)
+  {
+    if (rc1[r] != rc2[r])
+    {
+      if (diff_index >= 0)
+      {
+        // differ at more than one index
+        Trace("q-ext-rewrite-debug")
+            << "...more than one diff component." << std::endl;
+        return -1;
+      }
+      diff_index = r;
+    }
+  }
+  Assert(diff_index >= 0);
+  if (!rc1[diff_index].isConst() || !rc2[diff_index].isConst())
+  {
+    Trace("q-ext-rewrite-debug")
+        << "...non-constant diff components." << std::endl;
+    return -1;
+  }
+  // insert prefix
+  if (diff_index > 0)
+  {
+    nv.insert(nv.end(), rc1.begin(), rc1.begin() + diff_index);
+  }
+  Assert(rc1[diff_index] != rc2[diff_index]);
+  Node c1 = rc1[diff_index];
+  Node c2 = rc2[diff_index];
+  // do they differ by exactly one bit?
+  int bit_diff_index = -1;
+  unsigned csize = bv::utils::getSize(c1);
+  for (unsigned i = 0; i < csize; i++)
+  {
+    if (bv::utils::getBit(c1, i) != bv::utils::getBit(c2, i))
+    {
+      if (bit_diff_index >= 0)
+      {
+        // differ by more than one bit
+        nv.clear();
+        Trace("q-ext-rewrite-debug")
+            << "...more than one bit diff." << std::endl;
+        return -1;
+      }
+      bit_diff_index = i;
+    }
+  }
+  if (bit_diff_index >= 0)
+  {
+    std::vector<Node> split;
+    if (bit_diff_index + 1 < static_cast<int>(csize))
```

Comment:
This is way ugly. I'd suggest to use `unsigned` for `bit_diff_index` and use a flag that indicates if it has been set.